src/backend/distributed/utils/citus_stat_tenants.c

```diff
@@ -281,12 +286,26 @@ AttributeTask(char *tenantId, int colocationId, CmdType commandType)
 
 	MultiTenantMonitor *monitor = GetMultiTenantMonitor();
 	bool found = false;
+
+	/* Acquire the lock in shared mode to check if the tenant is already in the hash table. */
+	LWLockAcquire(&monitor->lock, LW_SHARED);
+
 	hash_search(monitor->tenants, &key, HASH_FIND, &found);
 
+	LWLockRelease(&monitor->lock);
+
 	/* If the tenant is not found in the hash table, we will track the query with a probability of StatTenantsSampleRateForNewTenants. */
 	if (!found)
 	{
-		int randomValue = rand() % 100;
+#if (PG_VERSION_NUM >= PG_VERSION_15)
+		double randomValue = pg_prng_double(&pg_global_prng_state);
+#else
+	    /* Set a good seed for random */
+    	srandom((unsigned int) time(NULL));
+
+		/* Generate a random double between 0 and 1 */
+		double randomValue = (double) random() / RAND_MAX;
```

Comment:
Should be MAX_RANDOM_VALUE
```suggestion
		double randomValue = (double) random() / MAX_RANDOM_VALUE;
```

From the PG source code comment on MAX_RANDOM_VALUE:

> CAUTION: Think not to replace this with RAND_MAX.  RAND_MAX defines the maximum value of the older rand() function, which is often different from --- and considerably inferior to --- random().